### PR TITLE
feat(compute): Regional and encrypted disk samples

### PR DESCRIPTION
**Pull request by Shnatsel**
- Add samples for:
    - Creating and deleting regional disks
    - Creating encrypted disks
    - Cloning disks
- Change some commented-out placeholder values for consistency

### CODE REVIEW

File: compute/disks/clone_encrypted_disk.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	// projectID := "your_project_id"
+	// zone := "us-west3-b"
+	// diskName := "your_disk_name"
+	// diskType := "zones/us-west3/diskTypes/pd-ssd"
+	// diskSizeGb := 120
+	// diskLink := "projects/your_project_id/global/disks/disk_name"
+	// encryptionKey := "SGVsbG8gZnJvbSBHb29nbGUgQ2xvdWQgUGxhdGZvcm0=" // in base64
+
+	ctx := context.Background()
+	disksClient, err := compute.NewDisksRESTClient(ctx)
```
Comment (telpirion) [inline compute/disks/clone_encrypted_disk.go:46]: Does this require the REST client? Or can you use the gRPC client?

File: compute/disks/clone_kms_encrypted_disk.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	// projectID := "your_project_id"
+	// zone := "us-west3-b"
+	// diskName := "your_disk_name"
+	// diskType := "zones/us-west3/diskTypes/pd-ssd"
+	// diskSizeGb := 120
+	// diskLink := "projects/your_project_id/global/disks/disk_name"
+	// kmsKeyLink := "projects/your_kms_project_id/locations/us-central1/keyRings/your_key_ring/cryptoKeys/your_key"
+
+	ctx := context.Background()
+	disksClient, err := compute.NewDisksRESTClient(ctx)
```
Comment (telpirion) [inline compute/disks/clone_kms_encrypted_disk.go:46]: Same as above, can you use a gRPC client?

**Review by telpirion — CHANGES_REQUESTED** [review]
This is a pretty large review (>500 LoC)--would you please split in half? Thank you!

File: compute/disks/clone_encrypted_disk.go (hunk 1 above)
Comment (Shnatsel) [reply compute/disks/clone_encrypted_disk.go:46]: No, Compute Engine doesn't have a gRPC client. More details here: https://github.com/GoogleCloudPlatform/golang-samples/pull/2780/files#r1044725617

**Comment by telpirion** [comment]
> While the raw line count is large, the files are mostly identical due to the large amount of boilerplate that Go requires. I believe it would be easier to review in one PR, since the files are so similar. Also, any changes requested in one of them would likely apply to all of them.
> 
> I can split it up if you insist, but I believe it to be counter-productive due to the repetitive nature of this code.

Fair enough. I can't promise that I'll get to them very quickly, though, given the size.

File: compute/disks/create_disk_from_snapshot.go [hunk 3]
```diff
@@ -29,12 +29,14 @@ import (
 func createDiskFromSnapshot(
 	w io.Writer,
 	projectID, zone, diskName, diskType, snapshotLink string,
+	diskSizeGb int64,
 ) error {
 	// projectID := "your_project_id"
-	// zone := "europe-central2-b"
+	// zone := "us-west3-b"
```
Comment (telpirion) [inline compute/disks/create_disk_from_snapshot.go:35]: Do developers have to use `us-west3-b`? If not, I'd make sure that this is indicated in the comments.

File: compute/disks/create_empty_disk.go [hunk 4]
```diff
@@ -29,11 +29,13 @@ import (
 func createEmptyDisk(
 	w io.Writer,
 	projectID, zone, diskName, diskType string,
+	diskSizeGb int64,
 ) error {
 	// projectID := "your_project_id"
-	// zone := "europe-central2-b"
+	// zone := "us-west3-b"
```
Comment (telpirion) [inline compute/disks/create_empty_disk.go:35]: Same as above.

**Comment by Shnatsel** [comment]
That's alright. FWIW `diff`ing the samples may help simplify the review, since most of the code is identical across the samples.

File: compute/disks/create_encrypted_disk.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+			Type:   proto.String(diskType),
+			SizeGb: proto.Int64(diskSizeGb),
+			DiskEncryptionKey: &computepb.CustomerEncryptionKey{
+				RawKey: &encryptionKey,
+			},
+		},
+	}
+
+	// if a source disk, image or snapshot has been specified, apply it
+	if diskLink != "" {
```
Comment (telpirion) [inline compute/disks/create_encrypted_disk.go:69]: Are these fields mutually exclusive?

File: compute/disks/create_kms_encrypted_disk.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+// createKmsEncryptedDisk creates a new disk in a project in given zone.
+// If you do not provide diskLink or imageLink, an empty disk will be created.
+func createKmsEncryptedDisk(
+	w io.Writer,
+	projectID, zone, diskName, diskType string,
+	diskSizeGb int64,
+	kmsKeyLink, imageLink, diskLink, snapshotLink string,
+) error {
+	// projectID := "your_project_id"
+	// zone := "us-west3-b"
```
Comment (telpirion) [inline compute/disks/create_kms_encrypted_disk.go:37]: Same comment as above.

File: compute/disks/create_kms_encrypted_disk.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+			Type:   proto.String(diskType),
+			SizeGb: proto.Int64(diskSizeGb),
+			DiskEncryptionKey: &computepb.CustomerEncryptionKey{
+				KmsKeyName: &kmsKeyLink,
+			},
+		},
+	}
+
+	// if a source disk, image or snapshot has been specified, apply it
+	if diskLink != "" {
```
Comment (telpirion) [inline compute/disks/create_kms_encrypted_disk.go:69]: Same as above: are these fields mutually exclusive?

File: compute/disks/create_regional.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+// createRegionalDisk creates a new empty regional disk
+func createRegionalDisk(
+	w io.Writer,
+	projectID, region string, replicaZones []string,
+	diskName, diskType string,
+	diskSizeGb int64,
+) error {
+	// projectID := "your_project_id"
+	// region := "us-west3"
```
Comment (telpirion) [inline compute/disks/create_regional.go:36]: Same as above.

File: compute/disks/create_regional_from_disk.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+// createRegionalDiskFromDisk creates a new regional disk with the contents of
+// an already existitng zonal disk. Disk type and size may differ.
+func createRegionalDiskFromDisk(
+	w io.Writer,
+	projectID, region string, replicaZones []string,
+	diskName, diskType, sourceDiskLink string,
+	diskSizeGb int64,
+) error {
+	// projectID := "your_project_id"
+	// region := "us-west3"
```
Comment (telpirion) [inline compute/disks/create_regional_from_disk.go:37]: Same as above.

File: compute/disks/create_regional_from_snapshot.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+// createRegionalDiskFromSnapshot creates a new regional disk with the contents of
+// an already existitng zonal disk snapshot.
+func createRegionalDiskFromSnapshot(
+	w io.Writer,
+	projectID, region string, replicaZones []string,
+	diskName, diskType, snapshotLink string,
+	diskSizeGb int64,
+) error {
+	// projectID := "your_project_id"
+	// region := "us-west3"
```
Comment (telpirion) [inline compute/disks/create_regional_from_snapshot.go:37]: Same as above.

File: compute/disks/create_regional_from_snapshot.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	// Exactly two replica zones must be specified
+	replicaZoneURLs := []string{
+		fmt.Sprintf("projects/%s/zones/%s", projectID, replicaZones[0]),
+		fmt.Sprintf("projects/%s/zones/%s", projectID, replicaZones[1]),
+	}
+
+	ctx := context.Background()
+	disksClient, err := compute.NewRegionDisksRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewRegionDisksRESTClient: %v", err)
```
Comment (telpirion) [inline compute/disks/create_regional_from_snapshot.go:53]: NIT: The current best practice is to use `%w` when passing errors along.

File: compute/disks/delete_disk.go [hunk 12]
```diff
@@ -27,7 +27,7 @@ import (
 // deleteDisk deletes a disk from a project.
 func deleteDisk(w io.Writer, projectID, zone, diskName string) error {
 	// projectID := "your_project_id"
-	// zone := "europe-central2-b"
+	// zone := "us-west3-b"
```
Comment (telpirion) [inline compute/disks/delete_disk.go:30]: See above.

File: compute/disks/delete_regional_disk.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+// deleteRegionalDisk permanently deletes a regional disk.
+func deleteRegionalDisk(w io.Writer, projectID, region, diskName string) error {
+	// projectID := "your_project_id"
+	// region := "us-west3"
```
Comment (telpirion) [inline compute/disks/delete_regional_disk.go:30]: Same as above.

File: compute/disks/disks_test.go [hunk 14]
```diff
@@ -156,18 +135,23 @@ func TestComputeDisksSnippets(t *testing.T) {
 
 	buf := &bytes.Buffer{}
 
-	t.Run("createDiskSnapshot and deleteDisk", func(t *testing.T) {
-		err := createDisk(ctx, tc.ProjectID, zone, diskName, sourceImage)
+	t.Run("Create zonal and regional disks from a snapshot and delete them", func(t *testing.T) {
```
Comment (telpirion) [inline compute/disks/disks_test.go:138]: Does this test need to be run concurrently with the following tests? If not, I'd break these into separate test cases. The `T.Run()` method is intended for testing concurrency.

File: compute/disks/disks_test.go [hunk 15]
```diff
@@ -184,28 +168,90 @@ func TestComputeDisksSnippets(t *testing.T) {
...
 
-	t.Run("createEmptyDisk", func(t *testing.T) {
+	t.Run("create, clone and delete an encrypted disk", func(t *testing.T) {
 		buf.Reset()
-		want = "Disk created"
+		want := "Disk created"
 
-		if err := createEmptyDisk(buf, tc.ProjectID, zone, diskName, diskType); err != nil {
-			t.Fatalf("createEmptyDisk got err: %v", err)
+		if err := createEncryptedDisk(buf, tc.ProjectID, zone, diskName, diskType, 20, "SGVsbG8gZnJvbSBHb29nbGUgQ2xvdWQgUGxhdGZvcm0=", "", "", ""); err != nil {
```
Comment (telpirion) [inline compute/disks/disks_test.go:240]: This set up step can be moved into `TestMain(m *testing.M)`.

**Review by telpirion — CHANGES_REQUESTED** [review]

File: compute/disks/disks_test.go (hunk 15 above)
Comment (Shnatsel) [reply compute/disks/disks_test.go:240]: This is not setup - this is the test body! It is calling a code sample wrapped into a function and ensuring that it works.

File: compute/disks/create_disk_from_snapshot.go (hunk 3 above)
Comment (Shnatsel) [reply compute/disks/create_disk_from_snapshot.go]: Done

File: compute/disks/create_regional_from_snapshot.go (hunk 11 above)
Comment (Shnatsel) [reply compute/disks/create_regional_from_snapshot.go:53]: All of these samples use `%v`, as do the rest of compute samples. How many of them do you want me to refactor? All disk samples, all compute samples, all Go samples? Do you want this to happen in this PR?

File: compute/disks/create_encrypted_disk.go (hunk 5 above)
Comment (Shnatsel) [reply compute/disks/create_encrypted_disk.go:70]: Yes, but unfortunately Go's type system is too primitive to express that. I've added a comment.

File: compute/disks/create_regional_from_snapshot.go (hunk 11 above)
Comment (telpirion) [reply compute/disks/create_regional_from_snapshot.go:53]: Let's switch over to using `%w` starting with this PR. We can refactor other samples on a "best efforts" basis.

File: compute/disks/disks_test.go (hunk 14 above)
Comment (telpirion) [reply compute/disks/disks_test.go:138]: I'm afraid the correctness of tests goes hand-in-hand with publishing samples. See the [Samples Guide](https://googlecloudplatform.github.io/samples-style-guide/#testing).

Please refactor into separate test cases.

**Review by telpirion — CHANGES_REQUESTED** [review]

**Comment by Shnatsel** [comment]
The samples guide specifies only the following about tests:

> Code snippets should have reasonable test coverage and all critical code paths should have integration tests that test against the production service.

Which is clearly fulfilled by the tests.

I believe the change you requested is misguided, but I'm going to go ahead and apply it.

**Review by telpirion — APPROVED** [review]
I'm still not a fan of the usage of `t.Run()` in these samples, but I won't stop you from merging them.

**Pull request merged**